### PR TITLE
Various performance optimizations / experiments [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
Puship WIP here to facilitate benchmarking.

### CODE REVIEW

**Comment by retronym** [comment]
🤞, the commits of this PR will start appearing as points in https://scala-ci.typesafe.com/grafana/dashboard/db/scala-benchmark?var-branch=2.12.x&var-branch=%3Cnone%3E&var-host=kalles@kalles@&from=1479937436145&to=1490002682574&var-hostId=kalles@kalles@ over the coming day or two.

**Comment by rorygraves** [comment]
Cool stuff

**Comment by retronym** [comment]
We're having some [problems](https://github.com/scala/scala-dev/issues/338) getting stable benchmark results on the new server, so I'm running the benchmarks on a spare machine of mine.

Results are appearing in the [wip-branch](https://scala-ci.typesafe.com/grafana/dashboard/db/wip-branch?from=1489990950397&to=1489991272204&var-branch=%3Cnone%3E&var-host=JasonMBP2014.local@JasonMBP2014.local@&var-source=All) dashboard. It will take about 8 hours to finish generating the results.

File: src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala [hunk 1]
```diff
@@ -298,19 +298,29 @@ abstract class BTypes {
...
-      case SHORT  => "S"
-      case INT    => "I"
-      case FLOAT  => "F"
-      case LONG   => "J"
-      case DOUBLE => "D"
-      case ClassBType(internalName) => "L" + internalName + ";"
-      case ArrayBType(component)    => "[" + component
-      case MethodBType(args, res)   => "(" + args.mkString + ")" + res
+    final override def toString: String = {
+      val builder = new StringBuilder(64)
```
Comment (rorygraves) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala]: You can save an extra allocations here by using the java StringBuilder directly.

Comment (rorygraves) [reply src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala]: Ignore me I think you did the change in another commit.

**Comment by retronym** [comment]
I haven't managed to get the benchmark server producing stable enough results, so I performed benchmarking locally. [Results](https://docs.google.com/spreadsheets/d/1qKX7XLoBnYRmRrCNXp5bNZC5gzQRUpXM2c96GkUKsJ8/edit?usp=sharing)

Hot compilation of `scalap` and `better-files` through each commit in this PR shows an aggregate performance improvement of 25-30%. 

image

The benefits will vary somewhat depending on the codebase being compiled, e.g. @rorygraves reports a 22% improvement for compiling `akka-actor`.

There are a few commits that appear to be clear winners, and then many that are hard to reason about in isolation as the individual performance changes are around the same level as the noise in the benchmark.

I'm going to close this pull request an submit smaller PRs, starting with changes that have the best cost/benefit ratio (easy to review patches / big performance improvement).

**Pull request closed without merging**